File: desktop/macos/Sources/VoiceBridgeAI/Settings/TranscriptPreferences.swift
import Foundation

/// Subtitle transcript storage path and filename preferences.
@MainActor
enum TranscriptPreferences {
    private static let directoryKey = "transcriptStorageDirectory"
    private static let filenameTemplateKey = "transcriptFilenameTemplate"
    private static let filePrefixKey = "transcriptFilePrefix"
    private static let fileExtensionKey = "transcriptFileExtension"
    private static let contentLayoutKey = "transcriptContentLayout"
    /// Kept from overlay toggle key for upgrade compatibility.
    private static let recordEnabledKey = "overlayRecordTranslations"

    static let filenameTokenHelp = "可用占位符：{prefix} {date} {time} {datetime} {mode}"

    /// Whether finalized subtitles are written to `storageDirectory` while a session runs.
    static var recordEnabled: Bool {
        get {
            if UserDefaults.standard.object(forKey: recordEnabledKey) == nil {
                return true
            }
            return UserDefaults.standard.bool(forKey: recordEnabledKey)
        }
        set {
            UserDefaults.standard.set(newValue, forKey: recordEnabledKey)
        }
    }

    static var defaultDirectory: URL {
        CloudProviderPreferences.dataDirectory().appendingPathComponent("transcripts", isDirectory: true)
    }

    static var storageDirectory: URL {
        get {
            let raw = UserDefaults.standard.string(forKey: directoryKey)?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            if raw.isEmpty { return defaultDirectory }
            return URL(fileURLWithPath: raw, isDirectory: true)
        }
        set {
            UserDefaults.standard.set(newValue.path, forKey: directoryKey)
        }
    }

    static var usesDefaultDirectory: Bool {
        UserDefaults.standard.string(forKey: directoryKey)?
            .trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true
    }

    static func resetDirectoryToDefault() {
        UserDefaults.standard.removeObject(forKey: directoryKey)
    }

    static var directoryDisplayPath: String {
        if usesDefaultDirectory {
            return "默认：\(defaultDirectory.path)"
        }
        return storageDirectory.path
    }

    /// 文件名模板，默认 `{prefix}_{datetime}`。
    static var filenameTemplate: String {
        get {
            let stored = UserDefaults.standard.string(forKey: filenameTemplateKey)?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return stored.isEmpty ? "{prefix}_{datetime}" : stored
        }
        set {
            let trimmed = newValue.trimmingCharacters(in: .whitespacesAndNewlines)
            UserDefaults.standard.set(trimmed.isEmpty ? "{prefix}_{datetime}" : trimmed, forKey: filenameTemplateKey)
        }
    }

    static var filePrefix: String {
        get {
            let stored = UserDefaults.standard.string(forKey: filePrefixKey)?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return stored.isEmpty ? "字幕记录" : stored
        }
        set {
            let trimmed = newValue.trimmingCharacters(in: .whitespacesAndNewlines)
            UserDefaults.standard.set(trimmed.isEmpty ? "字幕记录" : trimmed, forKey: filePrefixKey)
        }
    }

    enum FileFormat: String, CaseIterable {
        case markdown
        case plainText

        var fileExtension: String {
            switch self {
            case .markdown: return "md"
            case .plainText: return "txt"
            }
        }

        var label: String {
            switch self {
            case .markdown: return "Markdown (.md)"
            case .plainText: return "纯文本 (.txt)"
            }
        }
    }

    static var fileFormat: FileFormat {
        get {
            guard let raw = UserDefaults.standard.string(forKey: fileExtensionKey) else { return .markdown }
            return FileFormat(rawValue: raw) ?? .markdown
        }
        set {
            UserDefaults.standard.set(newValue.rawValue, forKey: fileExtensionKey)
        }
    }

    static var contentLayout: TranscriptContentLayout {
        get {
            guard let raw = UserDefaults.standard.string(forKey: contentLayoutKey),
                  let layout = TranscriptContentLayout(rawValue: raw) else {
                return .bilingualBlocks
            }
            return layout
        }
        set {
            UserDefaults.standard.set(newValue.rawValue, forKey: contentLayoutKey)
        }
    }

    static func sessionFilename(started: Date, modeId: String) -> String {
        let template = filenameTemplate
        let prefix = sanitizePathComponent(filePrefix)
        let mode = sanitizePathComponent(modeId)

        let dateFormatter = DateFormatter()
        dateFormatter.locale = Locale(identifier: "zh_CN")
        dateFormatter.dateFormat = "yyyy-MM-dd"
        let date = dateFormatter.string(from: started)

        dateFormatter.dateFormat = "HH-mm-ss"
        let time = dateFormatter.string(from: started)

        dateFormatter.dateFormat = "yyyy-MM-dd_HH-mm-ss"
        let datetime = dateFormatter.string(from: started)

        var name = template
            .replacingOccurrences(of: "{prefix}", with: prefix)
            .replacingOccurrences(of: "{date}", with: date)
            .replacingOccurrences(of: "{time}", with: time)
            .replacingOccurrences(of: "{datetime}", with: datetime)
            .replacingOccurrences(of: "{mode}", with: mode)

        name = sanitizePathComponent(name)
        if name.isEmpty { name = "\(prefix)_\(datetime)" }

        let ext = fileFormat.fileExtension
        if !name.lowercased().hasSuffix(".\(ext)") {
            name += ".\(ext)"
        }
        return name
    }

    static func sanitizePathComponent(_ value: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "" }
        let invalid = CharacterSet(charactersIn: "/:\\?%*|\"<>\n\r\t")
        let scalars = trimmed.unicodeScalars.map { invalid.contains($0) ? "_" : Character($0) }
        return String(scalars).trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: desktop/windows/VoiceBridgeAI/VoiceBridgeAI/Session/TranscriptPreferences.cs
using System.Globalization;
using System.Text.Json;

namespace VoiceBridgeAI.Session;

public static class TranscriptPreferences
{
    public const string FilenameTokenHelp = "可用占位符：{prefix} {date} {time} {datetime} {mode}";

    private const string FileName = "transcript-prefs.json";
    private static readonly object Gate = new();
    private static TranscriptPrefsData? _cache;

    public static event Action? Changed;

    public enum FileFormat
    {
        Markdown,
        PlainText,
    }

    public enum ContentLayout
    {
        BilingualBlocks,
        Combined,
        EnglishOnly,
        ChineseOnly,
    }

    public static bool RecordEnabled
    {
        get => Load().RecordEnabled ?? true;
        set
        {
            if (Load().RecordEnabled == value)
            {
                return;
            }

            Load().RecordEnabled = value;
            Persist();
            Changed?.Invoke();
        }
    }

    public static string DefaultDirectory =>
        Path.Combine(AppSupport.DataDirectory, "transcripts");

    public static string StorageDirectory
    {
        get
        {
            var raw = Load().StorageDirectory?.Trim() ?? "";
            return string.IsNullOrEmpty(raw) ? DefaultDirectory : raw;
        }
        set
        {
            Load().StorageDirectory = value.Trim();
            Persist();
            Changed?.Invoke();
        }
    }

    public static bool UsesDefaultDirectory =>
        string.IsNullOrWhiteSpace(Load().StorageDirectory);

    public static void ResetDirectoryToDefault()
    {
        Load().StorageDirectory = null;
        Persist();
        Changed?.Invoke();
    }

    public static string DirectoryDisplayPath =>
        UsesDefaultDirectory ? $"默认：{DefaultDirectory}" : StorageDirectory;

    public static string FilenameTemplate
    {
        get
        {
            var stored = Load().FilenameTemplate?.Trim() ?? "";
            return string.IsNullOrEmpty(stored) ? "{prefix}_{datetime}" : stored;
        }
        set
        {
            var trimmed = value.Trim();
            Load().FilenameTemplate = string.IsNullOrEmpty(trimmed) ? "{prefix}_{datetime}" : trimmed;
            Persist();
            Changed?.Invoke();
        }
    }

    public static string FilePrefix
    {
        get
        {
            var stored = Load().FilePrefix?.Trim() ?? "";
            return string.IsNullOrEmpty(stored) ? "字幕记录" : stored;
        }
        set
        {
            var trimmed = value.Trim();
            Load().FilePrefix = string.IsNullOrEmpty(trimmed) ? "字幕记录" : trimmed;
            Persist();
            Changed?.Invoke();
        }
    }

    public static FileFormat FileFormatValue
    {
        get
        {
            var raw = Load().FileExtension?.Trim();
            if (string.IsNullOrEmpty(raw))
            {
                return FileFormat.Markdown;
            }

            return raw.Equals("plainText", StringComparison.OrdinalIgnoreCase)
                ? FileFormat.PlainText
                : FileFormat.Markdown;
        }
        set
        {
            Load().FileExtension = value == FileFormat.PlainText ? "plainText" : "markdown";
            Persist();
            Changed?.Invoke();
        }
    }

    public static ContentLayout ContentLayoutValue
    {
        get
        {
            var raw = Load().ContentLayout?.Trim();
            if (string.IsNullOrEmpty(raw)
                || !Enum.TryParse<ContentLayout>(raw, ignoreCase: true, out var layout))
            {
                return ContentLayout.BilingualBlocks;
            }

            return layout;
        }
        set
        {
            Load().ContentLayout = value.ToString();
            Persist();
            Changed?.Invoke();
        }
    }

    public static string SessionFilename(DateTime started, string modeId)
    {
        var template = FilenameTemplate;
        var prefix = SanitizePathComponent(FilePrefix);
        var mode = SanitizePathComponent(modeId);
        var culture = CultureInfo.GetCultureInfo("zh-CN");

        var date = started.ToString("yyyy-MM-dd", culture);
        var time = started.ToString("HH-mm-ss", culture);
        var datetime = started.ToString("yyyy-MM-dd_HH-mm-ss", culture);

        var name = template
            .Replace("{prefix}", prefix, StringComparison.Ordinal)
            .Replace("{date}", date, StringComparison.Ordinal)
            .Replace("{time}", time, StringComparison.Ordinal)
            .Replace("{datetime}", datetime, StringComparison.Ordinal)
            .Replace("{mode}", mode, StringComparison.Ordinal);

        name = SanitizePathComponent(name);
        if (string.IsNullOrEmpty(name))
        {
            name = $"{prefix}_{datetime}";
        }

        var ext = FileExtensionFor(FileFormatValue);
        if (!name.EndsWith("." + ext, StringComparison.OrdinalIgnoreCase))
        {
            name += "." + ext;
        }

        return name;
    }

    public static string SanitizePathComponent(string value)
    {
        var trimmed = value.Trim();
        if (string.IsNullOrEmpty(trimmed))
        {
            return "";
        }

        var invalid = new HashSet<char>(Path.GetInvalidFileNameChars())
        {
            '/', '\\', '?', '%', '*', '|', '"', '<', '>', '\n', '\r', '\t',
        };

        var chars = trimmed.Select(ch => invalid.Contains(ch) ? '_' : ch).ToArray();
        return new string(chars).Trim();
    }

    public static string LabelFor(FileFormat format) => format switch
    {
        FileFormat.Markdown => "Markdown (.md)",
        FileFormat.PlainText => "纯文本 (.txt)",
        _ => format.ToString(),
    };

    public static string LabelFor(ContentLayout layout) => layout switch
    {
        ContentLayout.BilingualBlocks => "中英对照（分区）",
        ContentLayout.Combined => "中英结合（连续）",
        ContentLayout.EnglishOnly => "纯英文",
        ContentLayout.ChineseOnly => "纯中文",
        _ => layout.ToString(),
    };

    public static string ExportSuffixFor(ContentLayout layout) => layout switch
    {
        ContentLayout.BilingualBlocks => "_dual",
        ContentLayout.Combined => "_mix",
        ContentLayout.EnglishOnly => "_en",
        ContentLayout.ChineseOnly => "_zh",
        _ => "",
    };

    public static string FileExtensionFor(FileFormat format) => format switch
    {
        FileFormat.Markdown => "md",
        FileFormat.PlainText => "txt",
        _ => "md",
    };

    private static TranscriptPrefsData Load()
    {
        lock (Gate)
        {
            return _cache ??= ReadFromDisk();
        }
    }

    private static void Persist()
    {
        lock (Gate)
        {
            if (_cache is null)
            {
                return;
            }

            WriteToDisk(_cache);
        }
    }

    private static TranscriptPrefsData ReadFromDisk()
    {
        try
        {
            Directory.CreateDirectory(AppSupport.DataDirectory);
            var path = PrefsPath();
            if (!File.Exists(path))
            {
                return new TranscriptPrefsData();
            }

            var json = File.ReadAllText(path);
            return JsonSerializer.Deserialize<TranscriptPrefsData>(json) ?? new TranscriptPrefsData();
        }
        catch
        {
            return new TranscriptPrefsData();
        }
    }

    private static void WriteToDisk(TranscriptPrefsData data)
    {
        try
        {
            Directory.CreateDirectory(AppSupport.DataDirectory);
            File.WriteAllText(PrefsPath(), JsonSerializer.Serialize(data));
        }
        catch
        {
            // Preference persistence failure should not crash the app.
        }
    }

    private static string PrefsPath() => Path.Combine(AppSupport.DataDirectory, FileName);

    private sealed class TranscriptPrefsData
    {
        public string? StorageDirectory { get; set; }
        public string? FilenameTemplate { get; set; }
        public string? FilePrefix { get; set; }
        public string? FileExtension { get; set; }
        public string? ContentLayout { get; set; }
        public bool? RecordEnabled { get; set; }
    }
}
